### ISSUE TITLE
Expose public API for defining custom repository types

### ISSUE THREAD

**Issue by bmuschko** [gradle/gradle#1400]
### Expected Behavior

Gradle provides a public API for defining custom repository types. Typical example for custom repository types are P2, Ruby Gem repositories etc. Optionally, allow for `RepositoryHandler` to register a shortcut for the repository like `maven` and `ivy`.

### Current Behavior

At the moment a user would have to modify Gradle core (with tricks) to register a custom repository type. Shortcut methods on `RepositoryHandler` could be added by registering an extra method.

### Context

The use of Gradle in other domains like GoLang, Ruby etc. requires the use of types of repositories other than Maven or Ivy. A related discussion of this topic on the gradle-dev list:

http://gradle.1045684.n5.nabble.com/Adding-support-for-p2-repositories-td5713110.html

**Comment by adammurdoch** [comment]
I think we should separate this into 2 pieces: 

- Offer an API for build logic to reuse the Gradle transports, along with the goodness they offer: authentication, proxying, caching, offline support, instrumentation in build scans and IDEs, progress logging, parallel safety, etc.
- Offer APIs to allow build logic to contribute piece of logic to assemble into a repository. For example, logic that knows which locations to search for a given artifact, or logic that knows how to determine the available versions of a given library, or logic that knows how to calculate the meta-data of a component version, or ...

**Comment by nimish** [comment]
This would be great. Currently {py}gradle has to use a hacky importer that converts PyPI metadata into Ivy metadata. https://github.com/Ullink/gradle-repositories-plugin relies on monkeypatching the RepositoryHandler and intercepting method calls.

**Comment by tonit** [comment]
+1

**Comment by tobias-jenkner** [comment]
+1

**Comment by ysb33r** [comment]
Please take into account what was written in #2423.

**Comment by mkobit** [comment]
Making the dependency system, repositories, and cache usable from plugins would be useful to avoid reinventing the wheel in certain kinds of plugins.

**Comment by filippor** [comment]
Any news? 
For using p2 dependency im using  
`project.afterEvaluate` to add file dependency resolved and downloaded by tycho library
is there a better way?
look at  resolveDeps() method on https://github.com/filippor/testTychoGradle/blob/test-kotlin/testTychoGradle/buildSrc/src/main/kotlin/tycho/GradleTycho.kt

**Comment by noel-yap** [comment]
I'm thinking this feature could also be useful to be able to create a Github Release Asset repository (eg to access assets in https://github.com/nilslice/protolock/releases).

**Comment by mkobit** [comment]
@noel-yap that would definitely be helpful, I don't believe it can done today using the current APIs (see https://github.com/gradle/gradle/issues/5322 )

**Comment by Frontrider** [comment]
Any progress on this?

**Comment by melix** [comment]
No progress on this one. It's not the simplest unfortunately.

**Comment by Frontrider** [comment]
Yeah, took a deeper look into it. The dependency management is really hard coded.

**Comment by ferozco** [comment]
Has there been any progress on this? I think this is necessary for supporting other languages in a first class way

**Comment by jdiazcano** [comment]
Is there any possibility of describing the kind of (high level) changes that are needed? I am very much in the need of using my own repository (based on the maven repository but with a few changes). I have downloaded the project and tried making some changes but it is being very hard for me as everything seems using the "Default" or "Base" classes.

Since this is not having any traction, maybe someone else is willing to start this work! (In any case I'm taking some time to learn a little bit more about the internals of gradle and if I have enough time, work on it)

**Comment by mkobit** [comment]
It would definitely be great to see some movement here, either from Gradle itself or something that the community can help with. It is extremely frustrating that Gradle still doesn't have straightforward support for downloading a remote archive and using it in the build.

For a few examples of what I would consider other attempts to support other types of dependencies:

1. [Gradle source dependencies](https://blog.gradle.org/introducing-source-dependencies) have no documentation and are difficult to work with when they are non-Gradle projects (for example, https://github.com/gradle/gradle/issues/16752). I don't see any `@Incubating` annotations, so I don't know if they are considered stable. 
2. Gradle internally has recently added [toolchains for JVM projects](https://docs.gradle.org/current/userguide/toolchains.html), which _feel_ like another type of dependency. 
3. Kotlin has introduced NPM dependencies in their Kotlin JS ecosystem. Their support is deep. Below is an example of usage from [here](https://play.kotlinlang.org/hands-on/Building%20Web%20Applications%20with%20React%20and%20Kotlin%20JS/02_Setting_up):

    ```kotlin
    dependencies {
        //React, React DOM + Wrappers (chapter 3)
        implementation("org.jetbrains:kotlin-react:17.0.2-pre.154-kotlin-1.5.0")
        implementation("org.jetbrains:kotlin-react-dom:17.0.2-pre.154-kotlin-1.5.0")
        implementation(npm("react", "17.0.2"))
        implementation(npm("react-dom", "17.0.2"))
    
        //Kotlin Styled (chapter 3)
        implementation("org.jetbrains:kotlin-styled:5.2.3-pre.154-kotlin-1.5.0")
        implementation(npm("styled-components", "~5.2.3"))
    
        //Video Player (chapter 7)
        implementation(npm("react-youtube-lite", "1.0.1"))
    
        //Share Buttons (chapter 7)
        implementation(npm("react-share", "~4.2.1"))
    
        //Coroutines (chapter 8)
        implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.5.0-RC")
    }
    ```

-----

In contrast, Bazel largely already has different dependency types solved. From their [external dependencies documentation](https://docs.bazel.build/versions/master/external.html), with support for:

> * [Dependencies on other Bazel projects](https://docs.bazel.build/versions/master/external.html#bazel-projects)
> * [Dependencies on non-Bazel projects](https://docs.bazel.build/versions/master/external.html#non-bazel-projects)
> * [Dependencies on external packages](https://docs.bazel.build/versions/master/external.html#external-packages)

Are there any plans to break this apart, possibly building support for custom dependencies or at least remote archive dependencies?

**Comment by Fleshgrinder** [comment]
> […] for downloading a remote archive and using it in the build.

But that's actually a different kind of issue because an archive from some website is not a standardized kind of repository like asked for in this issue. This is really about things like https://rubygems.org/, https://crates.io/, etc. that have a stable and clear API like m2 repositories and for which it is possible to provide a stable repository handler.

For archives the best solution is the existing `IvyRepository` with its patterns. This has some limitations (no `HEAD` support) but those are not going to be lifted according to Gradle staff (as you know, since the original issue is from you 😉 https://github.com/gradle/gradle/issues/5322).

That said, providing an alternative to `IvyRepository` with support for custom patterns that can actually download things without `HEAD` is something that should be possible if this issue is resolved in a very flexible way. So maybe that is what you are asking for?

**Comment by ferozco** [comment]
The approach that we took with [gradle-typescript](https://github.com/gradlets/gradle-typescript) was to use an `IvyRepository`, point it at a [local proxy server](https://github.com/gradlets/gradle-typescript/blob/27c8b51999d56cce0b038a252c439bb49d3a5cfb/gradle-typescript/src/main/java/com/gradlets/gradle/typescript/shim/NpmArtifactoryShim.java), and have that proxy server convert between ivy and npm (ex. package.json -> ivy.descriptor). This allows us to use configuration resolution to resolve and download the "classpath" of a typescript project and then leverage a `TransformAction` to [unpack the npm bundles](https://github.com/gradlets/gradle-typescript/blob/27c8b51999d56cce0b038a252c439bb49d3a5cfb/gradle-typescript/src/main/java/com/gradlets/gradle/npm/NpmArtifactTransformAction.java). We have found this approach to work relatively well but would prefer to have a first class way of registering custom repositories

**Comment by filippor** [comment]
The approach I  use on https://github.com/filippor/p2Gradle is to take advantage that a dependency cam be a task with ConfigurableFileCollection output so a dependency can be declared with a function that create a task
example: 
dependencies {
   api p2.bundles('org.eclipse.core.resources:[3.13,3.14)')
}
